### ISSUE TITLE
asigning feature definitions for an OnDiskMSnExp object fails

### ISSUE THREAD

**Issue by SiggiSmara** [sneumann/xcms#263]
to reproduce the error:
generate an OnDiskMSnExp object with features (obnamex) and run the code:
```
allfeatures <- featureDefinitions(obnamex)
featureDefinitions(obnamex) <- allfeatures
```
It will fail with the error message: 

> Error in validObject(object) : 
>   invalid class "MsFeatureData" object: Some of the indices in column 'peakidx' of element 'featureDefinitions' do not match rows of the 'chromPeaks' matrix!

Not sure what is going on there, manually doing the check in validateMsFeatureData is not failing.

**Comment by jotsetung** [comment]
Hm, have problems reproducing. The code below works.

```r
library(xcms)
faahko_3_files <- c(system.file('cdf/KO/ko15.CDF', package = "faahKO"),
                    system.file('cdf/KO/ko16.CDF', package = "faahKO"),
                    system.file('cdf/KO/ko18.CDF', package = "faahKO"))
od_x <- readMSData(faahko_3_files, mode = "onDisk")
xod_x <- findChromPeaks(od_x, param = CentWaveParam(noise = 10000,
                                                              snthresh = 40))
xod_xg <- groupChromPeaks(xod_x, param = PeakDensityParam(
                   sampleGroups = rep(1, length(fileNames(xod_x)))))
xod_xgr <- adjustRtime(xod_xg, param = PeakGroupsParam(span = 0.4))
xod_xgrg <- groupChromPeaks(xod_xgr, param = PeakDensityParam(
                     sampleGroups = rep(1, length(fileNames(xod_x)))))


allf <- featureDefinitions(xod_xgrg)
featureDefinitions(xod_xgrg) <- allf
```

Did you do anything with `obnamex`, or did you somehow manipulate `allfeatures`? I assume `hasChromPeaks(obnamex)` returns `TRUE`.

**Comment by jotsetung** [comment]
Update:
```r
tmp <- fillChromPeaks(xod_xgrg)
allf <- featureDefinitions(tmp)
featureDefinitions(tmp) <- allf
Error in validObject(object) : 
  invalid class “MsFeatureData” object: Some of the indices in column 'peakidx' of element 'featureDefinitions' do not match rows of the 'chromPeaks' matrix!
```

**Comment by jotsetung** [comment]
OK, I think I know what's going on:

After filling in missing peaks we have
```r
nrow(chromPeaks(tmp))
[1] 272

```
peaks. `featureDefinitions<-` removes first already present feature definitions with the call `object <- dropFeatureDefinitions(object)`. `dropFeatureDefinition` is a clever function that wants to ensure that the data remains consistent. This means that, since the filled-in peaks are dependent on the feature definitions (or were generated based on them), it also removes the filled-in peaks along with the feature definitions:
```r
nrow(chromPeaks(dropFeatureDefinitions(tmp)))
[1] 248
```
This is actually really important, because, imagine you call `fillChromPeaks` several times, because you were not happy with the results, without removing previously filled-in peaks we would end in a large number of orphan peaks.

So, I would like to keep this behavior to ensure data consistency. In your case you could still use the replacement function, but before filling in missing peaks.

**Comment by SiggiSmara** [comment]
It is true that this is a good behavior if you are replacing the feature definitions with completely new definitions. But there is at least one other use case. In my instance, and I think here I am not alone, it is actually to sub set the current features, which will more often than not happen after calling fill peaks. Maybe this needs to be a new feature?

**Comment by jotsetung** [comment]
@sneumann yes. there needs to be a better check and error.

@SiggiSmara I'm somewhat not entirely happy with replacing the `featureDefinitions`. This brakes the concept of reproducibility in the `XCMSnExp` object, i.e. that you could repeat the analysis using the internally stored parameter classes and get to the same result. 
I will check how we could implement this properly. I could also think of a `filterFeatures` function/method to do that.

**Comment by SiggiSmara** [comment]
@jotsetung I agree that this breaks reproducibility. What is the original use case for implementing the function? I can only think of an instance where you basically want to manually integrate over multiple regions and you assign a feature definition with no chrompeaks ids or my use case of filtering the already defined features.

**Comment by SiggiSmara** [comment]
Just in case anyone else runs into the same problem before we decide on what to do, the obvious workaround is:
```R
library(xcms)
faahko_3_files <- c(system.file('cdf/KO/ko15.CDF', package = "faahKO"),
                    system.file('cdf/KO/ko16.CDF', package = "faahKO"),
                    system.file('cdf/KO/ko18.CDF', package = "faahKO"))
od_x <- readMSData(faahko_3_files, mode = "onDisk")
xod_x <- findChromPeaks(od_x, param = CentWaveParam(noise = 10000,
                                                              snthresh = 40))
xod_xg <- groupChromPeaks(xod_x, param = PeakDensityParam(
                   sampleGroups = rep(1, length(fileNames(xod_x)))))
xod_xgr <- adjustRtime(xod_xg, param = PeakGroupsParam(span = 0.4))
xod_xgrg <- groupChromPeaks(xod_xgr, param = PeakDensityParam(
                     sampleGroups = rep(1, length(fileNames(xod_x)))))


allfBefore <- featureDefinitions(xod_xgrg)
tmp <- fillChromPeaks(xod_xgrg)
allf <- featureDefinitions(tmp)
#obviously some more fancy filtering than this in real life
rowids <- 1:10 
featureDefinitions(xod_xgrg) <- allfBefore[rowids,]
xod_xgrg <- fillChromPeaks(xod_xgrg)
allfAfter <- featureDefinitions(xod_xgrg)
all(dim(allfAfter)==c(10,9))
#result: TRUE
#continue with more work on the filtered features
```

**Comment by SiggiSmara** [comment]
@jotsetung After sleeping on it I think that if a user filters or assign new feature definitions I think it might be a good idea to issue a warning that this will break reproducibility. 

Plus, I don't think it will be easy to implement a filter function on the feature definitions. There are too many variables that you might want to use to filter on. In theory there are already a lot of settings that you use in the steps to get to the feature definitions in the first place. If you need added filtering after then it is likely that this is something unique to the data you are working with.

The best idea I can think of  right now, and it isn't all that good, is having one more column in the data table (with a heading such as 'active') and update a TRUE/FALSE value based on user input of rowids. Any call to featureDefinitions() would then always filter on this column before returning a data frame to the user and then there has to be an override / setting to give the entire data frame to the user.

**Comment by jotsetung** [comment]
Re original use case of the function: it was designed to be used internally by e.g. the `fillChromPeaks` method to update the feature definitions. I never really thought this to be used by the user (well, but somehow exported the method - my fault).

I'm against adding a new column in the data. I think that would make things more complicated.

What exactly is it what you want to do? Do you want to sub-set the existing `featureDefinitions` by some means? For that I still think a `filterFeatureDefinitions` (that takes a `logical` or `integer` to select the rows to keep + your suggestion to show a warning) should be OK.

**Comment by SiggiSmara** [comment]
Subsetting the feature definitions is what I set out to do when I ran into this error/feature. Then of course I start theorizing how to use a function in creative ways :). 

What you suggest would solve my use case perfectly fine. The idea behind adding a new column was to keep a path back to reproducibility open but it is also fine just to make users aware that they are potentially closing that path.
